notebooks/byod.py

```diff
@@ -0,0 +1,224 @@
+            return set()
+
+    def join_keyed_rows(keyed_rows_a: Dict[str, Any], keyed_rows_b: Dict[str, Any]) -> Dict[str, Any]:
+        assert not keyed_row_columns(keyed_rows_a).intersection(keyed_row_columns(keyed_rows_b))
+        common_keys = set(keyed_rows_a.keys()).intersection(set(keyed_rows_b.keys()))
+        return {k: dict(**keyed_rows_a[k], **keyed_rows_b[k])
+                for k in common_keys}
+
+    def join_data_tables(new_table: str, tables_to_join: list, join_column: str):
+        keyed_rows = get_keyed_rows(tables_to_join[0], join_column)
+        for table_name in tables_to_join[1:]:
+            keyed_rows = join_keyed_rows(keyed_rows, get_keyed_rows(table_name, join_column))
+        upload_rows(new_table, [{join_column: k, **row} for k, row in keyed_rows.items()])
+
+    def create_cram_crai_table(table: str, listing: Iterable[str]):
+        crams = dict()
+        crais = dict()
+        for key in listing:
+            _, filename = key.rsplit("/", 1)
+            sample, ext = filename.split(".", 1)
+            if "cram" == ext:
+                crams[sample] = key
+            elif "crai" == ext:
+                crais[sample] = key
+            else:
+                continue
+        samples = sorted(crams.keys())
+        upload_columns(table, dict(sample=samples,
+                                   cram=[crams[s] for s in samples],
+                                   crai=[crais[s] for s in samples]))
+
+with callysto.Cell("markdown"):
+    """
+    ## Install gsutil on your local computer using option 2 instructions in this [document](https://support.terra.bio/hc/en-us/articles/360024056512-Uploading-to-a-workspace-Google-bucket).
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ## Organize your upload to Terra
+    Using instructions from the document above, you will use the gsutil tool to upload data from your local computer to
+    your Terra workspace. To effectively use the example here, we suggest you upload the data with the suggestions
+    below.
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ### Find the path to this workspace bucket
+    """
+
+with callysto.Cell("markdown"):
+    """
+    Using the os package, you can print your workspace bucket path.
+    """
+
+with callysto.Cell("python"):
+    bucket = os.environ["WORKSPACE_BUCKET"]
+    print(bucket)
+
+with callysto.Cell("markdown"):
+    """
+    ### Add a prefix to your bucket path to organize your data
+    In this example, we add the prefix 'test-crai-cram'. In the terminal of your computer, you will call something like:
+
+    `gsutil cp /Users/Documents/Example.cram gs://your_bucket_info/test-crai-cram/`
+    """
+
+with callysto.Cell("markdown"):
+    """
+    ## Preview the data in your workspace bucket
+    """
+
+with callysto.Cell("python"):
+    #!gsutil ls {bucket}
+    pass
+
+with callysto.Cell("markdown"):
+    """
+    # Generate a data table that links to the data in your workspace bucket
```

Comment:
Were you planning to add example code here for using your functions above to create a table?

Reply:
Yes, will add.

Reply:
example added